### PR TITLE
Initial JSONRPC interface

### PR DESCRIPTION
**Pull request by mrallen1**
Problem to solve: We want to expose a JSONRPC interface into miner/validators

### CODE REVIEW

File: config/sys.config [hunk 1]
```diff
@@ -36,6 +36,7 @@
   ]},
  {blockchain,
   [
+   {jsonrpc_port, 4467},
```
Comment (Vagabond) [inline config/sys.config:39]: Should this not be off by default?

File: rebar.config [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
 {deps, [
         {blockchain, {git, "https://github.com/helium/blockchain-core.git",
-                      {branch, "master"}}},
+                      {branch, "andymck/check-nonce-during-isvalid"}}},
```
Comment (Vagabond) [inline rebar.config:13]: I believe this merged?

File: config/sys.config (hunk 1 above)
Comment (madninja) [reply config/sys.config:39]: should this not be _the_ way to interact with miner?

Comment (evanmcc) [reply config/sys.config:39]: I'm with @madninja

File: rebar.config (hunk 2 above)
Comment (evanmcc) [reply rebar.config:13]: yes, this should be based on master

File: config/sys.config (hunk 1 above)
Comment (mrallen1) [reply config/sys.config:39]: I think we should turn it _on_ by default but only bind to localhost. That seems like the least surprising and "secure" default to me. Other opinions are welcome, though :)

File: rebar.config (hunk 2 above)
Comment (mrallen1) [reply rebar.config:13]: ah yeah you're right

File: src/jsonrpc/miner_jsonrpc_accounts.erl [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                                    #{ dc_balance => 0,
+                                       dc_nonce => 0
+                                     }
+                            end
+                    end,
+    lists:foldl(fun(Fun, Map) ->
+                        maps:merge(Map, Fun())
+                end,
+                #{
+                  address => ?BIN_TO_B58(Address)
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_accounts.erl:55]: if you're using TO_B58 everywhere else you may as well do it here?

File: src/jsonrpc/miner_jsonrpc_info.erl [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            GWLoc = blockchain_ledger_gateway_v2:location(Gateway),
+            GWOwnAddr = libp2p_crypto:pubkey_bin_to_p2p(
+                blockchain_ledger_gateway_v2:owner_address(Gateway)
+            ),
+            iolist_to_binary([
+                lists:concat(["Gateway location: ", GWLoc]),
+                lists:concat(["Owner address: ", GWOwnAddr])
+            ]);
+        _ ->
+            "gateway not found in ledger"
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_info.erl:113]: return `undefined` or `null` here.. easier to deal with in client code

File: src/jsonrpc/miner_jsonrpc_info.erl [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    Macs.
+
+get_firmware_version() ->
+    iolist_to_binary(os:cmd("cat /etc/lsb_release")).
+
+get_uptime() ->
+    {UpTimeMS, _} = statistics(wall_clock),
+    UpTimeSec = (UpTimeMS div 1000) rem 1000000,
+    DaysTime = calendar:seconds_to_daystime(UpTimeSec),
+    format_days_time(DaysTime).
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_info.erl:98]: just return this in seconds.. normalized, and easier to parse for clients than a string.

File: src/jsonrpc/miner_jsonrpc_snapshot.erl [hunk 6]
```diff
@@ -0,0 +1,39 @@
+-module(miner_jsonrpc_snapshot).
+
+-include("miner_jsonrpc.hrl").
+
+-behavior(miner_jsonrpc_handler).
+
+-define(MAX_LOOKBACK, 25). %% only search the previous 25 blocks
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_snapshot.erl:7]: unused?

File: src/jsonrpc/miner_jsonrpc_txn.erl [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        #{ <<"result">> => B64 }
+    catch
+        T:E:St ->
+            lager:error("Couldn't do add gateway via JSONRPC because: ~p ~p: ~p",
+                        [T, E, St]),
+            Error = io_lib:format("~p", [E]),
+            #{ <<"error">> => Error }
+    end;
+handle_rpc(<<"txn_assert_location">>, #{ owner := OwnerB58,
+                                         location := Loc } = Params) ->
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_txn.erl:48]: use `lat` and `lon` arguments instead of string splitting

File: test/miner_txn_mgr_SUITE.erl [hunk 8]
```diff
@@ -274,77 +274,6 @@ txn_invalid_nonce_test(Config) ->
 
     ok.
 
-txn_out_of_sequence_nonce_max_block_spantest(Config) ->
```
Comment (madninja) [inline test/miner_txn_mgr_SUITE.erl:277]: how are these changes related to this PR?

Comment (mrallen1) [reply test/miner_txn_mgr_SUITE.erl:277]: They were needed to boot blockchains for testing, otherwise totally unrelated

File: src/jsonrpc/miner_jsonrpc_accounts.erl (hunk 3 above)
Comment (mrallen1) [reply src/jsonrpc/miner_jsonrpc_accounts.erl:55]: Yeah I decided I should consolidate macros into the miner_jsonrpc.hrl file but then scattered around some other files are the imports from blockchain_utils.hrl .. so I will clean up those refs to be consistent

File: src/jsonrpc/miner_jsonrpc_snapshot.erl (hunk 6 above)
Comment (mrallen1) [reply src/jsonrpc/miner_jsonrpc_snapshot.erl:7]: Copypasta mistake that wasn't cleaned up; removed

File: src/jsonrpc/miner_jsonrpc_info.erl [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            GWLoc = blockchain_ledger_gateway_v2:location(Gateway),
+            GWOwnAddr = libp2p_crypto:pubkey_bin_to_p2p(
+                blockchain_ledger_gateway_v2:owner_address(Gateway)
+            ),
+            iolist_to_binary([
+                lists:concat(["Gateway location: ", GWLoc]),
+                lists:concat(["Owner address: ", GWOwnAddr])
+            ]);
+        _ ->
+            <<"undefined">>
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_info.erl:112]: I meant the atom undefined so it shows up as null in the resulting json

File: src/jsonrpc/miner_jsonrpc_info.erl [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+get_gateway_info(Chain, PubKey) ->
+    Ledger = blockchain:ledger(Chain),
+    case blockchain_ledger_v1:find_gateway_info(PubKey, Ledger) of
+        {ok, Gateway} ->
+            GWLoc = blockchain_ledger_gateway_v2:location(Gateway),
+            GWOwnAddr = libp2p_crypto:pubkey_bin_to_p2p(
+                blockchain_ledger_gateway_v2:owner_address(Gateway)
+            ),
+            iolist_to_binary([
+                lists:concat(["Gateway location: ", GWLoc]),
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_info.erl:108]: that doesn't look like it'll translate to json well at all?

File: test/miner_txn_mgr_SUITE.erl (hunk 8 above)
Comment (madninja) [reply test/miner_txn_mgr_SUITE.erl:277]: so they were not used by txn_mgr test code?

File: src/jsonrpc/miner_jsonrpc_txn.erl (hunk 7 above)
Comment (mrallen1) [reply src/jsonrpc/miner_jsonrpc_txn.erl:48]: Yeah good ideas

File: src/jsonrpc/miner_jsonrpc_txn.erl [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+parse_location(#{ <<"h3">> := H3 }) ->
+    try
+        h3:from_string(H3),
+        {ok, H3}
+    catch
+        _:_ ->
+            {error, {invalid_location, H3}}
+    end;
+parse_location(#{ <<"lat">> := LatIn,
+                  <<"long">> := LongIn }) ->
```
Comment (madninja) [inline src/jsonrpc/miner_jsonrpc_txn.erl:81]: it's `lon` not long

Comment (mrallen1) [reply src/jsonrpc/miner_jsonrpc_txn.erl:81]: ok

Comment (mrallen1) [reply src/jsonrpc/miner_jsonrpc_txn.erl:81]: h3 literally expects a string not a binary string... so

**Review by madninja — APPROVED** [review]
